File: DesignPatterns/01_GOF23/DP_examples/C/03_Behavioral/TemplateMethod/abstract_class.c
#include "abstract_class.h"

// 初始化抽象类
void BeverageMaker_init(BeverageMaker* self) {
    self->makeBeverage = BeverageMaker_makeBeverage;
    self->customerWantsCondiments = BeverageMaker_customerWantsCondiments;
}

// 模板方法：制作饮料的流程
void BeverageMaker_makeBeverage(BeverageMaker* self) {
    BeverageMaker_boilWater();
    if (self->brew) {
        self->brew(self);
    }
    BeverageMaker_pourInCup();
    if (self->customerWantsCondiments && self->customerWantsCondiments(self)) {
        if (self->addCondiments) {
            self->addCondiments(self);
        }
    }
}

// 具体方法：烧水
void BeverageMaker_boilWater() {
    printf("Boiling water\n");
}

// 具体方法：倒入杯子
void BeverageMaker_pourInCup() {
    printf("Pouring into cup\n");
}

// 钩子方法：顾客是否需要调料
int BeverageMaker_customerWantsCondiments(BeverageMaker* self) {
    return 1;
}

File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/TemplateMethod/abstract_class.go
package template_method

import "fmt"

// BeverageMaker 饮料制作抽象类
type BeverageMaker interface {
	MakeBeverage()
	Brew()
	AddCondiments()
	CustomerWantsCondiments() bool
}

// AbstractBeverageMaker 抽象实现
type AbstractBeverageMaker struct {
	BeverageMaker
}

// MakeBeverage 模板方法：制作饮料的流程
func (ab *AbstractBeverageMaker) MakeBeverage() {
	ab.boilWater()
	ab.Brew()
	ab.pourInCup()
	if ab.CustomerWantsCondiments() {
		ab.AddCondiments()
	}
}

// boilWater 具体方法：烧水
func (ab *AbstractBeverageMaker) boilWater() {
	fmt.Println("Boiling water")
}

// pourInCup 具体方法：倒入杯子
func (ab *AbstractBeverageMaker) pourInCup() {
	fmt.Println("Pouring into cup")
}

// CustomerWantsCondiments 钩子方法：顾客是否需要调料
func (ab *AbstractBeverageMaker) CustomerWantsCondiments() bool {
	return true
}
